internal/store/store.go

```diff
@@ -89,235 +106,246 @@ func customCATransport(ca []byte) http.RoundTripper {
 
-	if baseURL := c.OpsManagerVersionManifestURL(); baseURL != "" {
-		s.baseURL = baseURL
+type CredentialsGetter interface {
+	PublicAPIKey() string
+	PrivateAPIKey() string
+}
+
+func WithAuthentication(c CredentialsGetter) Option {
+	return func(s *Store) error {
+		s.username = c.PublicAPIKey()
+		s.password = c.PrivateAPIKey()
+		return nil
 	}
+}
 
-	client, err := defaultClient(c)
-	if err != nil {
-		return nil, err
+func withAtlasClient(client *http.Client) Option {
+	return func(s *Store) error {
+		opts := []atlas.ClientOpt{atlas.SetUserAgent(userAgent)}
+		if s.baseURL != "" {
+			opts = append(opts, atlas.SetBaseURL(s.baseURL))
+		}
+		c, err := atlas.New(client, opts...)
+		if err != nil {
+			return err
+		}
+		s.client = c
+		return nil
 	}
+}
 
-	if err := s.setOpsManagerClient(client); err != nil {
-		return nil, err
+func withOpsManagerClient(client *http.Client) Option {
+	return func(s *Store) error {
+		opts := []opsmngr.ClientOpt{opsmngr.SetUserAgent(userAgent)}
+		if s.baseURL != "" {
+			opts = append(opts, opsmngr.SetBaseURL(s.baseURL))
+		}
+		c, err := opsmngr.New(client, opts...)
+		if err != nil {
+			return err
+		}
+
+		s.client = c
+		return nil
 	}
+}
 
-	return s, nil
+type TransportConfigGetter interface {
+	OpsManagerCACertificate() string
+	OpsManagerSkipVerify() string
 }
 
-// NewPrivateUnauth gets the appropriate client for the atlas private api
-func NewPrivateUnauth(c Config) (*Store, error) {
-	s := new(Store)
-	s.service = c.Service()
-	if s.service != config.CloudService {
-		return nil, fmt.Errorf("unsupported service: %s", s.service)
+func NetworkPresets(c TransportConfigGetter) Option {
+	options := make([]Option, 0)
+	if caCertificate := c.OpsManagerCACertificate(); caCertificate != "" {
+		options = append(options, WithCACertificate(caCertificate))
 	}
-	s.baseURL = atlas.CloudURL
+	if skipVerify := c.OpsManagerSkipVerify(); skipVerify != yes {
+		options = append(options, SkipVerify())
+	}
+	return Options(options...)
+}
+
+type Config interface {
+	CredentialsGetter
+	TransportConfigGetter
+	Service() string
+	OpsManagerURL() string
+}
 
+func PublicAuthenticatedPreset(c Config) Option {
+	options := []Option{Service(c.Service()), WithAuthentication(c)}
 	if configURL := c.OpsManagerURL(); configURL != "" {
-		s.baseURL = configURL
+		options = append(options, WithBaseURL(configURL), WithPublicPathBaseURL())
 	}
+	options = append(options, NetworkPresets(c))
+	return Options(options...)
+}
 
-	client, err := defaultClient(c)
-	if err != nil {
-		return nil, err
+func PublicUnauthenticatedPreset(c Config) Option {
+	options := []Option{Service(c.Service())}
+	if configURL := c.OpsManagerURL(); configURL != "" {
+		options = append(options, WithBaseURL(configURL), WithPublicPathBaseURL())
 	}
+	options = append(options, NetworkPresets(c))
+	return Options(options...)
+}
 
-	if err2 := s.setAtlasClient(client); err2 != nil {
-		return nil, err2
+func PrivateAuthenticatedPreset(c Config) Option {
```

Comment:
Authenticated private API

Reply:
I see why we are doing this but I also think that this way is much less intuitive.

Reply:
I can agree, it may take a while to get use to but this makes the `New` method now more maintainable and testable as we can now test the options without depending on a `Config`

This refactor is nothing more than a [dependency inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle) `New` no longer depends on `Config` which since the beginning was a bad call from my side, and I knew it but it kept things simple but as we added more `New` methods that choice started to show why it was a bad option, since `New` no longer depends on `Config` it's what makes it easier to test and maintain, we don't need `PrivateAuthenticatedPreset` I added the presets to keep usage simple but they can go any time and `New` will still work